### ISSUE TITLE
TypeError: edge.initializeClrFunc is not a function

### ISSUE THREAD

**Issue by vadi2** [agracio/electron-edge-js-quick-start#5]
This seems to be a popular one with edge.js. Running Ubuntu 17.10, node 8.11.1, and using `dotnet build` to get the dlls. 

Error log in Electron:
```
Uncaught TypeError: edge.initializeClrFunc is not a function
    at Object.exports.func (/home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/lib/edge.js:181:17)
    at Object.<anonymous> (/home/vadi/Programs/electron-edge-js-quick-start/renderer.js:10:34)
    at Object.<anonymous> (/home/vadi/Programs/electron-edge-js-quick-start/renderer.js:59:3)
    at Module._compile (module.js:569:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
    at Module.require (module.js:513:17)
    at require (internal/module.js:11:18)
```

Error in the console:

```
CoreClrEmbedding::Initialize - Failed to initialize CoreCLR, HRESULT: 0x80070057
```

Build log:

```
vadi@volga:~/Programs$ git clone https://github.com/agracio/electron-edge-js-quick-start.git
Cloning into 'electron-edge-js-quick-start'...
remote: Counting objects: 40, done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 40 (delta 3), reused 8 (delta 2), pack-reused 28
Unpacking objects: 100% (40/40), done.
vadi@volga:~/Programs$ cd electron-edge-js-quick-start/
vadi@volga:~/Programs/electron-edge-js-quick-start$ npm install

> edge-cs@1.2.1 install /home/vadi/Programs/electron-edge-js-quick-start/node_modules/edge-cs
> node tools/install.js

MSBUILD : error MSB1003: Specify a project or solution file. The current working directory does not contain a project or solution file.
Microsoft (R) Build Engine version 15.7.145.53551 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

MSBUILD : error MSB1003: Specify a project or solution file. The current working directory does not contain a project or solution file.

> electron-edge-js@8.3.2 install /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js
> node tools/install.js

make: Entering directory '/home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/build'
  TOUCH Release/obj.target/edge_nativeclr.stamp
  CXX(target) Release/obj.target/edge_coreclr/src/common/v8synchronizationcontext.o
  CXX(target) Release/obj.target/edge_coreclr/src/common/callbackhelper.o
  CXX(target) Release/obj.target/edge_coreclr/src/common/edge.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/coreclrembedding.o
../src/CoreCLREmbedding/coreclrembedding.cpp: In static member function ‘static HRESULT CoreClrEmbedding::Initialize(BOOL)’:
../src/CoreCLREmbedding/coreclrembedding.cpp:581:7: warning: variable ‘foundEdgeJs’ set but not used [-Wunused-but-set-variable]
  bool foundEdgeJs = false;
       ^~~~~~~~~~~
At global scope:
cc1plus: warning: unrecognized command line option ‘-Wno-unused-private-field’
cc1plus: warning: unrecognized command line option ‘-Wno-redundant-move’
cc1plus: warning: unrecognized command line option ‘-Wno-mismatched-tags’
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/coreclrfunc.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/coreclrnodejsfunc.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/coreclrfuncinvokecontext.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/coreclrnodejsfuncinvokecontext.o
  CXX(target) Release/obj.target/edge_coreclr/src/common/utils.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/pal/pal_utils.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/pal/trace.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/fxr/fx_ver.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/fxr/fx_muxer.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/json/casablanca/src/json/json.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/json/casablanca/src/json/json_parsing.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/json/casablanca/src/json/json_serialization.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/json/casablanca/src/utilities/asyncrt_utils.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/deps/deps_format.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/deps/deps_entry.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/deps/deps_resolver.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/host/args.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/host/coreclr.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/host/libhost.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/host/runtime_config.o
  CXX(target) Release/obj.target/edge_coreclr/src/CoreCLREmbedding/pal/pal.unix.o
  SOLINK_MODULE(target) Release/obj.target/edge_coreclr.node
  COPY Release/edge_coreclr.node
  ACTION binding_gyp_build_managed_target_restore_bootstrap_packages lib/bootstrap/project.lock.json
  Restoring packages for /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/src/double/Edge.js/Edge.js.csproj...
  Restoring packages for /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/lib/bootstrap/bootstrap.csproj...
  Generating MSBuild file /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/lib/bootstrap/obj/bootstrap.csproj.nuget.g.props.
  Generating MSBuild file /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/lib/bootstrap/obj/bootstrap.csproj.nuget.g.targets.
  Restore completed in 447.42 ms for /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/lib/bootstrap/bootstrap.csproj.
  Generating MSBuild file /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/src/double/Edge.js/obj/Edge.js.csproj.nuget.g.props.
  Generating MSBuild file /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/src/double/Edge.js/obj/Edge.js.csproj.nuget.g.targets.
  Restore completed in 503.34 ms for /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/src/double/Edge.js/Edge.js.csproj.
  ACTION binding_gyp_build_managed_target_compile_bootstrap lib/bootstrap/bin/Release/netcoreapp1.1/bootstrap.dll
Microsoft (R) Build Engine version 15.7.145.53551 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 49.4 ms for /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/lib/bootstrap/bootstrap.csproj.
  Restore completed in 46.29 ms for /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/src/double/Edge.js/Edge.js.csproj.
  You are working with a preview version of the .NET Core SDK. You can define the SDK version via a global.json file in the current project. More at https://go.microsoft.com/fwlink/?linkid=869452
  Edge.js -> /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/src/double/Edge.js/bin/Release/netcoreapp1.1/EdgeJs.dll
  bootstrap -> /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/lib/bootstrap/bin/Release/netcoreapp1.1/bootstrap.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.79
  TOUCH Release/obj.target/build_managed.stamp
make: Leaving directory '/home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/build'

> electron@1.8.4 postinstall /home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron
> node install.js

npm notice created a lockfile as package-lock.json. You should commit this file.
added 156 packages in 11.453s
vadi@volga:~/Programs/electron-edge-js-quick-start$ cd src/
vadi@volga:~/Programs/electron-edge-js-quick-start/src$ dotnet build
Microsoft (R) Build Engine version 15.7.145.53551 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restoring packages for /home/vadi/Programs/electron-edge-js-quick-start/src/ExternalLibrary/ExternalLibrary.csproj...
  Restoring packages for /home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/QuickStart.Core.csproj...
  Generating MSBuild file /home/vadi/Programs/electron-edge-js-quick-start/src/ExternalLibrary/obj/ExternalLibrary.csproj.nuget.g.props.
  Generating MSBuild file /home/vadi/Programs/electron-edge-js-quick-start/src/ExternalLibrary/obj/ExternalLibrary.csproj.nuget.g.targets.
  Restore completed in 151.48 ms for /home/vadi/Programs/electron-edge-js-quick-start/src/ExternalLibrary/ExternalLibrary.csproj.
  Generating MSBuild file /home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/obj/QuickStart.Core.csproj.nuget.g.props.
  Generating MSBuild file /home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/obj/QuickStart.Core.csproj.nuget.g.targets.
  Restore completed in 274.48 ms for /home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/QuickStart.Core.csproj.
  You are working with a preview version of the .NET Core SDK. You can define the SDK version via a global.json file in the current project. More at https://go.microsoft.com/fwlink/?linkid=869452
  You are working with a preview version of the .NET Core SDK. You can define the SDK version via a global.json file in the current project. More at https://go.microsoft.com/fwlink/?linkid=869452
  ExternalLibrary -> /home/vadi/Programs/electron-edge-js-quick-start/src/ExternalLibrary/bin/Debug/netcoreapp2.0/ExternalLibrary.dll
LocalMethods.cs(10,35): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread. [/home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/QuickStart.Core.csproj]
LocalMethods.cs(45,35): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread. [/home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/QuickStart.Core.csproj]
LocalMethods.cs(50,35): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread. [/home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/QuickStart.Core.csproj]
  QuickStart.Core -> /home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/bin/Debug/netcoreapp2.0/QuickStart.Core.dll

Build succeeded.

LocalMethods.cs(10,35): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread. [/home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/QuickStart.Core.csproj]
LocalMethods.cs(45,35): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread. [/home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/QuickStart.Core.csproj]
LocalMethods.cs(50,35): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread. [/home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/QuickStart.Core.csproj]
    3 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.04
vadi@volga:~/Programs/electron-edge-js-quick-start/src$ npm start

> electron-edge-js-quick-start@1.0.0 start /home/vadi/Programs/electron-edge-js-quick-start
> electron .

CoreClrEmbedding::Initialize - Failed to initialize CoreCLR, HRESULT: 0x80070057
```

**Comment by vadi2** [comment]
I'm not sure if this fixes it or just hides the error, but doing the following:

```js
const baseNetAppPath = path.join(__dirname, '/src/QuickStart.Core/bin/Debug/netcoreapp2.0');

process.env.EDGE_USE_CORECLR = 1;
process.env.EDGE_APP_ROOT = baseNetAppPath;

const edge = require('electron-edge-js');

var baseDll = path.join(baseNetAppPath, 'QuickStart.Core.dll');

var getAppDomainDirectory = edge.func({
    assemblyFile: baseDll,
    typeName: 'QuickStart.Core.LocalMethods',
    methodName: 'GetAppDomainDirectory'
});
```

now gives:

```
 JavaScript error occurred in the main process
Uncaught Exception:
Error: Could not load type 'QuickStart.Core.LocalMethods'
    at Object.exports.func (/home/vadi/Programs/electron-edge-js-quick-start/node_modules/electron-edge-js/lib/edge.js:181:17)
    at Object.<anonymous> (/home/vadi/Programs/electron-edge-js-quick-start/main.js:27:34)
```

**Comment by agracio** [comment]
Have you followed provided instructions? First do `npm install` and only then compile .sln? QuickStart.Core references dlls inside `electron-edge-js` folder and copies them to build output.

**Comment by vadi2** [comment]
Yes, I did `npm install` first - see log above. Here's how the folder with the dlls looks like:

```
$ ls /home/vadi/Programs/electron-edge-js-quick-start/src/QuickStart.Core/bin/Debug/netcoreapp2.0
EdgeJs.dll                         Microsoft.CodeAnalysis.CSharp.Workspaces.dll       Microsoft.DotNet.InternalAbstractions.dll  QuickStart.Core.dll                     System.Composition.Hosting.dll
EdgeJs.pdb                         Microsoft.CodeAnalysis.dll                         Microsoft.DotNet.PlatformAbstractions.dll  QuickStart.Core.pdb                     System.Composition.Runtime.dll
ExternalLibrary.dll                Microsoft.CodeAnalysis.VisualBasic.dll             Microsoft.Extensions.DependencyModel.dll   refs                                    System.Composition.TypedParts.dll
ExternalLibrary.pdb                Microsoft.CodeAnalysis.VisualBasic.Workspaces.dll  Newtonsoft.Json.dll                        System.Composition.AttributedModel.dll
Microsoft.CodeAnalysis.CSharp.dll  Microsoft.CodeAnalysis.Workspaces.dll              QuickStart.Core.deps.json                  System.Composition.Convention.dll
```

Is it correct?

**Comment by radencode** [comment]
I am getting the same error:

`TypeError: edge.initializeClrFunc is not a function
    at Object.exports.func (C:\Users\Jenster\Desktop\Development\reflow\reflow-client\current\release\Reflow-win32-x64\resources\app.asar\node_modules\electron-edge-js\lib\edge.js:182)
    at A.edgeInit (main.bundle.effe34fd3444d7d46fc7.js:1)
    at new A (main.bundle.effe34fd3444d7d46fc7.js:1)
    at new e (main.bundle.effe34fd3444d7d46fc7.js:1)
    at constructClassInstance (main.bundle.effe34fd3444d7d46fc7.js:1)
    at beginWork (main.bundle.effe34fd3444d7d46fc7.js:1)
    at n (main.bundle.effe34fd3444d7d46fc7.js:1)
    at o (main.bundle.effe34fd3444d7d46fc7.js:1)
    at i (main.bundle.effe34fd3444d7d46fc7.js:1)
    at Q (main.bundle.effe34fd3444d7d46fc7.js:1)`

Here are the versions that I am using:

`Electron: 2.0.2
Node: 8.9.3
electron-edge-js: 8.3.6
`

I am using the same configuration with `edge-js` and it works perfectly, however when I use `electron-edge-js` and build the app with `electron-builder` it gives me the above error. 

Did you guys find a fix for this? Am I doing something wrong? 
I would really appreciate your help.

**Comment by vadi2** [comment]
> Did you guys find a fix for this?


The issue is still there as you see, so I just created a command-line wrapper for my binary and just call it via Node.js's proces function to get the work done.

**Comment by agracio** [comment]
Quick start does not cover `electron-builder` created apps. When you package with electron builder it is likely that the paths specified in `process.env.EDGE_APP_ROOT` are no longer valid. Try using `console.log(process.env.EDGE_APP_ROOT)` to determine the problems with path.

**Comment by agracio** [comment]
@vadi2 `electron-edge-js` 8.3.6 has improvements for path handling, perhaps quick-start will work correctly on Linux. I have not had a chance yet to test it on Linux but it fixed issues on macOS. You can give it another try.

**Comment by agracio** [comment]
Update: just tested on Ununtu 18.04, runs without any problems.

**Comment by radencode** [comment]
@agracio `process.env.EDGE_APP_ROOT` is correct and I get the same error even if I don't use `electron-builder` it still gives me `edge.initializeClrFunc is not a function`

**Comment by agracio** [comment]
Ok so just be be clear on the issue:
You cloned a fresh copy of this repo and have not done any changes to it. Followed the steps from README in correct order and after running `npm start` command got this error?
I have test the latest version on Windows/macOS/Linux and did not encounter any problems so far.

**Comment by radencode** [comment]
I just re did everything and it works. I have no idea what the issue is but
it works now.

**Comment by radencode** [comment]
@agracio I figured out what my issue is: I am using `webpack `to bundle my `react.js` app and instead of `C:\Users\Jenster\Desktop\Development\reflow\reflow-client\electron-edge-js-quick-start\lib` the bundled path is `lib` Any idea how I can provide `process.env.EDGE_APP_ROOT` a relative path instead of absolute path?

**Comment by radencode** [comment]
@agracio I fixed it. I put the path in `node env variable` in my `electron main.js` file and then picked it up in my `react.js` component. Anyway, thank you for maintaining this project and providing support for .Net Core 2.0. You are awesome, keep up the good work. ;)

**Comment by grofit** [comment]
I just realised I am getting this issue on windows 10, here were my steps:

- Fork then clone repo (you could just clone it directly)
- run `npm i`
- run `dotnet build src/QuickStart.sln`
- run `npm start:standard` (get error)
- run `npm start` (get other error)

**Comment by iempeek** [comment]
I’m experiencing the same issue as @grofit
Same steps on Windows 10 x64.
require(edgeNative) in edge.js returns '{}'

**Comment by JonathanLydall** [comment]
I'm getting the exact same error, but I've found out something which seems to help a bit so I'm hoping it can lead someone else here to find a solution to the problem.

The Quickstart worked for me, but when I made a new project using Visual Studio 2017, it wouldn't work until I added `<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>` to my `.csproj` file, to make it look like:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Library</OutputType>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="EdgeJs" Version="9.3.4" />
  </ItemGroup>

</Project>
```

I'm still struggling though because that doesn't fix my existing project which has been running fine on Edge for a long time, but we're now trying to migrate it to .NET Core.

Will post if I work out more.

**Comment by ukenny** [comment]
After troubleshooting this for work I may have found a work-around. If I download `Release v2.2.4` with `SDK 2.2.203` I get this same error. But if I use a `global.json` with my .net core project to force it to use an older version of the SDK (and rebuild the .net core project) I can successfully run my project without issues. 

>{
>  "sdk": {
>    "version": "2.2.202"  }
>}

Make sure you download the older version of the SDK you're forcing the project to use!

The only different seems to be that they added C# 8.0 support in Release `2.2.4`. Can somebody who's had this issue try this and see if it fixes it for you?

**Comment by risikoruk** [comment]
Like @grofit and @IhorKrupin mentioned, those precise steps do not make the standard .NET version work.

The core one works okay for me though. With the standard instead, I immediately get: Uncaught TypeError: edge.initializeClrFunc is not a function

Also, why is the start:standard crossed over in the ReadMe? Is it no longer working?

**Comment by agracio** [comment]
Standard is crossed over because it is almost impossible to get it to work. To get to the bottom of the error try running https://github.com/agracio/edge-js-quick-start. Unlike Electron version it will display underlying error. The problem appears to be with dll versions for some of the references in .NET Standard project as some being imported from nuget with correct versions and some pulled in from your .NET Standard install folder that will have higher versions than requested by project. You might be able to solve it by changing the referenced versions in .csproj file.

**Comment by corylulu** [comment]
This is old, but I think I found the problem. After testing on 3 different systems and not being able to get it to initialize properly on Linux, I found that in **deps_entry.cpp**, the `EDGE_PLATFORM_NIX` was not being defined, and because of that, the nuget packages path resolutions weren't being lower cased. https://github.com/agracio/edge-js/pull/101

_Edit: Upon further investigation, I'm finding that the deps hash check just isn't valid. The .deps.json's "sha512" value simply doesn't match the "nkpkg.sha512" value on Linux... Should just be ignored if the versions match._
